### PR TITLE
Bug 2037209: refactor for AlibabaResourceReference changes

### PR DESCRIPTION
**Pull request by elmiko**
This change updates the logic around security group IDs, vswitch IDs,
and resource group IDs to match the union changes in the resource
reference API. AlibabaResourceReferences now contain a `Type` variable
which indicates whether it is an ID or a list of Tags.

The behavior has not been changed for security groups, vswitches, and
resource groups. This change is being added to help allow for scenarios
where the ResourceGroupID is not known and must be searched. Now that
the AlibabaResourceReference contains a discriminating union of ID and
Tags, we can properly address scenarios with limited information.

### CODE REVIEW

**Comment by elmiko** [comment]
this is most likely going to fail until we merge the API changes in https://github.com/openshift/api/pull/1096

File: pkg/actuators/machine/instances.go
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:459]: Is vSwitch an optional field? Do we need to check for a nil `VSwitch` before we start doing this?

Comment (elmiko) [reply pkg/actuators/machine/instances.go:459]: it says optional in the api, but that field is not a point to struct but an actual struct. i had _thought_ i tried doing a nil compare here and the compiler barked at me

File: pkg/actuators/machine/instances.go
Comment (elmiko) [reply pkg/actuators/machine/instances.go:828]: it seems like we fail if the resource group is nil currently, the way the code is structured i'm not sure how it's supposed to find the security groups without the group id. but maybe that's ok?

File: pkg/actuators/machine/instances.go
Comment (elmiko) [reply pkg/actuators/machine/instances.go:459]: looking through the code again, it seems like although VSwitchID is optional in the API, it would fail if you did not specify one or it could not determine which vswitch to use. i'm thinking we should change the API to make it not optional.

Comment (elmiko) [reply pkg/actuators/machine/instances.go:459]: these alibaba docs[0], indicate that a vswitch is required when using vpc mode, but that using classic mode does not require one. do we allow machine creation outside of a vpc?

if not, should we make `VpcID`[1] required?

[0] https://www.alibabacloud.com/help/en/doc-detail/63440.html
[1] https://github.com/openshift/api/blob/master/machine/v1/types_alibabaprovider.go#L85

File: pkg/actuators/machine/instances.go
Comment (elmiko) [reply pkg/actuators/machine/instances.go:828]: i asked around a little bit, and we should not allow creation of resources in the default group. the user should always specify the resource group.

File: pkg/actuators/machine/instances.go
Comment (elmiko) [reply pkg/actuators/machine/instances.go:459]: asked around and the assumption should be that we require VPC mode. which means that vswitch must be supplied as well.

**Comment by elmiko** [comment]
updated to match the API changes and to reflect our requirements

File: pkg/actuators/machine/instances.go
Comment (JoelSpeed) [reply pkg/actuators/machine/instances.go:828]: I saw a comment about most customers using the default group, are we sure this is correct?

**Comment by elmiko** [comment]
@jianli-wei it looks like it is still unable to build this, until we get the api change in place.

**Comment by elmiko** [comment]
here is the installer patch that should work this as well, https://github.com/openshift/installer/pull/5562

**Comment by elmiko** [comment]
i have added a second commit to vendor from my branch of the api, this is only for testing and should be replaced before we merge.

**Comment by elmiko** [comment]
now that the API changes have merged, i have revendored this repository. we need one final round of manual testing before releasing the hold here.

**Review by alexander-demichev — APPROVED** [review]
/lgtm

File: pkg/actuators/machine/instances.go
Comment (elmiko) [reply pkg/actuators/machine/instances.go:828]: +1, we agreed that the user should always provide this information, even in the case where they want to use the default resource group. this will help ensure that the user isn't creating resources in the wrong group by accident.

**Comment by kwoodson** [comment]
Test with this code today:
```
        resourceGroup:
          id: rg-xxxxxx
          type: ID
```

and 
```
test-rx6ff-worker-us-east-1a-vdr6m   Ready    worker   14m   v1.23.0+06791f6
test-rx6ff-worker-us-east-1b-8l777   Ready    worker   14m   v1.23.0+06791f6
test-rx6ff-worker-us-east-1b-plcst   Ready    worker   16m   v1.23.0+06791f6
```

```
openshift-machine-api   test-rx6ff-worker-us-east-1a-vdr6m   Running   ecs.g6.large    us-east-1   us-east-1a   33m
openshift-machine-api   test-rx6ff-worker-us-east-1b-8l777   Running   ecs.g6.large    us-east-1   us-east-1b   33m
openshift-machine-api   test-rx6ff-worker-us-east-1b-plcst   Running   ecs.g6.large    us-east-1   us-east-1b   33m

```

/lgtm

**Comment by elmiko** [comment]
removing the hold here as we have a good test signal. thanks @kwoodson !
/hold cancel

**Pull request merged**
